File: apps/ios/Unfiled/Features/Capture/CaptureComposerRules.swift
import Foundation

/// What the composer allows and what it sends when the owner attached photos. Speaking a capture
/// is the keyboard's dictation key, which puts the words straight into the text.
enum CaptureComposerRules {
    static let maximumPhotos = 4
    static let maximumCharacters = 10_000

    /// Words or photos are enough to send; too many characters never are.
    static func canSend(content: String, attachmentCount: Int) -> Bool {
        guard content.utf16.count <= maximumCharacters else { return false }
        let words = content.trimmingCharacters(in: .whitespacesAndNewlines)
        return !words.isEmpty || attachmentCount > 0
    }

    /// The owner's words, or a plain placeholder when they typed nothing, so a capture always
    /// has text the organizer and the Inbox can show.
    static func rawContent(content: String, kinds: [LocalAttachmentKind]) -> String {
        let words = content.trimmingCharacters(in: .whitespacesAndNewlines)
        if !words.isEmpty { return words }
        switch kinds.filter({ $0 == .image }).count {
        case 0: return ""
        case 1: return "Photo"
        default: return "Photos"
        }
    }

    static func canAdd(_ kind: LocalAttachmentKind, to kinds: [LocalAttachmentKind]) -> Bool {
        switch kind {
        case .image: return kinds.filter { $0 == .image }.count < maximumPhotos
        case .audio: return false
        }
    }

    static func remainingPhotos(given kinds: [LocalAttachmentKind]) -> Int {
        max(0, maximumPhotos - kinds.filter { $0 == .image }.count)
    }
}

File: apps/web/src/lib/capture/capture-composer-rules.ts
import { MAX_CAPTURE_IMAGES } from "@unfiled/contracts";

/**
 * What the composer allows and what it sends when the owner attached photos. These are the
 * phone's `CaptureComposerRules`: a capture written on either client can be sent with words, with
 * photos, or with both, and a capture sent with photos alone still carries text the organizer and
 * the Inbox can show.
 */
export const MAX_CAPTURE_CHARACTERS = 10_000;
export const MAX_CAPTURE_PHOTOS = MAX_CAPTURE_IMAGES;

/** Words or photos are enough to send; too many characters never are. */
export function canSendCapture(content: string, photoCount: number): boolean {
  if (content.length > MAX_CAPTURE_CHARACTERS) return false;
  return content.trim().length > 0 || photoCount > 0;
}

/**
 * The owner's words, or a plain placeholder when they typed nothing, so a capture always has text
 * the organizer and the Inbox can show.
 */
export function captureRawContent(content: string, photoCount: number): string {
  const words = content.trim();
  if (words.length > 0) return words;
  if (photoCount === 1) return "Photo";
  return photoCount > 1 ? "Photos" : "";
}

/** How many more photos this capture can carry. */
export function remainingCapturePhotos(photoCount: number): number {
  return Math.max(0, MAX_CAPTURE_PHOTOS - photoCount);
}
